src/hotspot/share/opto/memnode.cpp

```diff
@@ -1829,6 +1829,574 @@ AllocateNode* LoadNode::is_new_object_mark_load() const {
+      (!Matcher::match_rule_supported(Op_ReverseBytesS) ||
+       !Matcher::match_rule_supported(Op_ReverseBytesI) ||
+       !Matcher::match_rule_supported(Op_ReverseBytesL))) {
+    // Reverse Bytes is not supported
+    return;
+  }
+
+  // All checks are passed
+  _last_op_index = last_op_index;
+  _order = order;
+
+  for (int i=0; i<collected; i++) {
+    merge_list.push(array[i]);
+  }
+#ifdef ASSERT
+  if (is_trace_basic()) {
+    tty->print_cr("[TraceMergeLoads]: dump final merge info list, collected: %d", collected);
+    for (int i=0; i < merge_list.length(); i++) {
+      MergeLoadInfo* info = merge_list.at(i);
+      info->dump();
+    }
+  }
+#endif
+  return;
+}
+
+// Make the merged load from list and replace the last Or oper
+LoadNode* MergePrimitiveLoads::make_merged_load(const MergeLoadInfoList& merge_list) {
+  if (merge_list.is_empty()) {
+    return nullptr;
+  }
+
+  assert(merge_list.length() >= 2 && _last_op_index >= 0 && _last_op_index < merge_list.length(), "sanity");
+  LoadNode* load = _order == LowToHigh ? merge_list.at(0)->_load : merge_list.at(merge_list.length()-1)->_load;
+  // Get address of merged load
+  if (_order == LowToHigh) {
+    load = merge_list.at(0)->_load;
+  } else {
+    load = merge_list.at(merge_list.length()-1)->_load;
+  }
+  Node* adr  = load->in(MemNode::Address);
+  Node* ctrl = load->in(MemNode::Control);
+  Node* mem  = load->in(MemNode::Memory);
+  Node* last_op = merge_list.at(_last_op_index)->_or;
+
+  const TypePtr* at = load->adr_type();
+  const Type* rt = nullptr;
+
+  int merge_size = merge_list.length() * _load->memory_size();
+  BasicType bt = T_ILLEGAL;
+  switch (merge_size) {
+    case 4: bt = T_INT;   rt = TypeInt::INT;   break;
+    case 8: bt = T_LONG;  rt = TypeLong::LONG; break;
+    case 2: // Not merged as LoadS
+    default: {
+      ShouldNotReachHere();
+      break;
+    }
+  }
+
+  LoadNode* merged_load = LoadNode::make(*_phase, ctrl, mem, adr,
+                                         at, rt, bt, MemNode::unordered)->isa_Load();
+
+  if (merged_load == nullptr) {
+    return nullptr;
+  }
+
+  merged_load->set_unaligned_access();
+  merged_load->set_mismatched_access();
+
+  Node* replace = merged_load;
+  if (_require_reverse_bytes) {
+    assert(_load->memory_size() == 1, "only implemented for bytes");
+    if (merge_size == 8) {
+      replace = _phase->transform(new ReverseBytesLNode(merged_load));
+    } else {
+      assert(merge_size == 4, "sanity");
+      replace = _phase->transform(new ReverseBytesINode(merged_load));
+    }
+    NOT_PRODUCT( if (is_trace_basic()) { tty->print("[TraceMergeLoads] new ReverseBytes node:"); replace->dump(); tty->cr(); })
+    _phase->is_IterGVN()->_worklist.push(replace);
+  }
+
+  assert(last_op != nullptr && (last_op->Opcode() == Op_OrI || last_op->Opcode() == Op_OrL), "sanity");
+  _phase->is_IterGVN()->replace_node(last_op, replace);
+  _phase->is_IterGVN()->_worklist.push(merged_load);
```

Comment:
If you did this in `OrNode::Ideal`, then you just have to return the new load, and `IGVN` takes care of the replacing. That is the code pattern we use everywhere else.

Reply:
The optimization has been moved to `OrNode::Ideal`